### PR TITLE
Change device_setter.py for choose ps device

### PR DESCRIPTION
**Pull request by HW-ZZ**
Add arguments "ps_device_name" for replica_device_setter(), in order to choose ps decive .
Reference the issue[#3762](https://github.com/tensorflow/tensorflow/issues/3762)

### CODE REVIEW

**Comment by mrry** [comment]
I'm not sure I see the point of this change. Can't you pass (e.g.) `"/job:ps/cpu:0"` to the existing `ps_device` argument if (as in #3762) you want to put all the variables on the CPU by default?

**Comment by HW-ZZ** [comment]
This is the code block in def replica_device_setter()
```
 ps_job_name = ps_device.lstrip("/job:")
 if ps_job_name not in cluster_spec or cluster_spec[ps_job_name] is None:
      return None
```
if the argument `ps_device="/job:ps/cpu:0"` it will return None

**Comment by mrry** [comment]
OK, that `lstrip()` is completely wrong. Thanks for pointing it out.

More properly, it should do:

```python
ps_job_name = pydev.DeviceSpec.from_string(ps_device).job
```

Do you want to try making that change in this PR?

**Comment by HW-ZZ** [comment]
@mrry  Yes, and thanks for your reply

**Comment by HW-ZZ** [comment]
@mrry I have changed it

**Comment by mrry** [comment]
Great, thanks! Can you also add a test that covers this case to `device_setter_test.py`?

I imagine you could copy `testPS2TasksWithDevice()` and make it `testPS2TasksWithCPUConstraint()`. The code would look something like:

```python
  def testPS2TasksWithCPUConstraint(self):
    cluster_spec = tf.train.ClusterSpec({
        "sun": ["sun0:2222", "sun1:2222", "sun2:2222"],
        "moon": ["moon0:2222", "moon1:2222"]})

    with tf.device(tf.train.replica_device_setter(
        ps_device="/job:moon/cpu:0", worker_device="/job:sun",
        cluster=cluster_spec.as_cluster_def())):
      v = tf.Variable([1, 2])
      w = tf.Variable([2, 1])
      a = v + w
      self.assertDeviceEqual("/job:moon/task:0/cpu:0", v.device)
      self.assertDeviceEqual("/job:moon/task:0/cpu:0", v.initializer.device)
      self.assertDeviceEqual("/job:moon/task:1/cpu:0", w.device)
      self.assertDeviceEqual("/job:moon/task:1/cpu:0", w.initializer.device)
      self.assertDeviceEqual("/job:sun", a.device)
```

**Comment by HW-ZZ** [comment]
@mrry  I have added testPS2TasksWithCPUConstraint() device_setter_test.py, any other test should be add?

File: tensorflow/python/training/device_setter_test.py [hunk 1]
```diff
@@ -77,6 +77,23 @@ def testPS2TasksWithDevice(self):
       self.assertDeviceEqual("/job:moon/task:1", w.device)
       self.assertDeviceEqual("/job:moon/task:1", w.initializer.device)
       self.assertDeviceEqual("/job:sun", a.device)
+  
+  def testPS2TasksWithCPUConstraint(self):
+     cluster_spec = tf.train.ClusterSpec({
+        "sun": ["sun0:2222", "sun1:2222", "sun2:2222"],
+        "moon": ["moon0:2222", "moon1:2222"]})
```
Comment (mrry) [inline tensorflow/python/training/device_setter_test.py]: It looks like the indentation is slightly inconsistent here: to satisfy our linter (which follows [these style rules](https://google.github.io/styleguide/pyguide.html)), the code should be indented two spaces at each block, and statement continuations (such as L83 and L84) should be indented four spaces.

Can you please fix that up? Then we should be good to go!

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Comment by mrry** [comment]
Looks good to me! Thanks for making this change.

**Comment by rmlarsen** [comment]
Thanks for the contribution!

**Pull request merged**